dataset/src/test/scala/frameless/forward/CubeTests.scala

```diff
@@ -0,0 +1,293 @@
+
+      val sumByAB = data.groupBy(x => (x.a, x.b))
+        .mapValues { xs => xs.map(_.c).sum }
+        .toVector.map { case ((a, b), c) => (a, b, c) }.sortBy(x => (x._1, x._2))
+
+      framelessSumByAB ?= sumByAB
+    }
+
+    check(forAll(prop[Byte, Int, Long] _))
+  }
+
+  test("cube('a).mapGroups(('a, toVector(('a, 'b))") {
+    def prop[
+    A: TypedEncoder,
+    B: TypedEncoder
+    ](data: Vector[X2[A, B]]): Prop = {
+      val dataset = TypedDataset.create(data)
+      val A = dataset.col[A]('a)
+
+      val datasetGrouped = dataset
+        .cube(A)
+        .deserialized.mapGroups((a, xs) => (a, xs.toVector))
+        .collect().run().toMap
+
+      val dataGrouped = data.groupBy(_.a)
+
+      datasetGrouped ?= dataGrouped
+    }
+
+    check(forAll(prop[Short, Option[Short]] _))
+    check(forAll(prop[Option[Short], Short] _))
+    check(forAll(prop[X1[Option[Short]], Short] _))
+  }
+
+  test("cube('a).flatMapGroups(('a, toVector(('a, 'b))") {
+    def prop[
+    A: TypedEncoder : Ordering,
+    B: TypedEncoder : Ordering
+    ](data: Vector[X2[A, B]]): Prop = {
+      val dataset = TypedDataset.create(data)
+      val A = dataset.col[A]('a)
+
+      val datasetGrouped = dataset
+        .cube(A)
+        .deserialized.flatMapGroups((a, xs) => xs.map(x => (a, x)))
+        .collect().run()
+        .sorted
+
+      val dataGrouped = data
+        .groupBy(_.a).toSeq
+        .flatMap { case (a, xs) => xs.map(x => (a, x)) }
+        .sorted
+
+      datasetGrouped ?= dataGrouped
+    }
+
+    check(forAll(prop[Short, Option[Short]] _))
+    check(forAll(prop[Option[Short], Short] _))
+    check(forAll(prop[X1[Option[Short]], Short] _))
+  }
+
+  test("cube('a, 'b).flatMapGroups((('a,'b) toVector((('a,'b), 'c))") {
+    def prop[
+    A: TypedEncoder : Ordering,
+    B: TypedEncoder : Ordering,
+    C: TypedEncoder : Ordering
+    ](data: Vector[X3[A, B, C]]): Prop = {
+      val dataset = TypedDataset.create(data)
+      val cA = dataset.col[A]('a)
+      val cB = dataset.col[B]('b)
+
+      val datasetGrouped = dataset
+        .cube(cA, cB)
+        .deserialized.flatMapGroups((a, xs) => xs.map(x => (a, x)))
+        .collect().run()
+        .sorted
+
+      val dataGrouped = data
+        .groupBy(t => (t.a,t.b)).toSeq
+        .flatMap { case (a, xs) => xs.map(x => (a, x)) }
+        .sorted
+
+      datasetGrouped ?= dataGrouped
+    }
+
+    check(forAll(prop[Short, Option[Short], Long] _))
+    check(forAll(prop[Option[Short], Short, Int] _))
+    check(forAll(prop[X1[Option[Short]], Short, Byte] _))
+  }
+
+  test("cubeMany('a).agg(sum('b))") {
+    def prop[A: TypedEncoder : Ordering, Out: TypedEncoder : Numeric]
+    (data: List[X1[A]])(implicit summable: CatalystSummable[A, Out]): Prop = {
+      val dataset = TypedDataset.create(data)
+      val A = dataset.col[A]('a)
+
+      val received = dataset.cubeMany(A).agg(count()).collect().run().toVector.sortBy(_._2)
```

Comment:
It returns `TypedDataset[Tuple1[Option[A]]]` instead of `TypedDataset[(Option[A], Long)]` for some reason.